### PR TITLE
Cupertino text selection menu customization

### LINKED ISSUE #24407: Allow more button options to the text selection toolbar
Flutter's Cupertino:
Empty Selection: Paste, Select All
Filled Selection: Cut, Copy Paste 

Native iOS Apps:
Empty Selection: Select, Select All, Paste, Insert Photo
Filled Selection: Cut, Copy, Paste, Replace, Arrow, bold/italicized/underlined, Look Up, Share, Arrow, Strikethrough, Insert Photo.

### LINKED ISSUE #31412: Add custom option to context menu.
When you select a piece of text, it has a "more info" (the three dots) button where apps register options (like "translate to rot13") and I was wondering if this was possible with Flutter? 

I couldn't quite key in what it would be called in the documentation.

### LINKED ISSUE #36985: Allow custom TextSelectionControls inside SelectableText
## Use case

Custom Menu for SelectableText

### PR DESCRIPTION
**Pull request by justinmc**
It's now possible to customize the buttons and the appearance of our iOS text selection menu.

Screen Shot 2021-01-08 at 9 52 22 AM

This follows my [previous work](https://github.com/flutter/flutter/pull/69428) to do the same thing for Material.  The API is as identical as possible.

The API for creating custom menus is still not ideal, though, because I limited these two PRs to non-breaking changes for now.  In the long term, I'd like to refactor TextSelectionControls (tracked in https://github.com/flutter/flutter/issues/73574).

I've created a separate [example repo](https://github.com/justinmc/flutter-text-selection-menu-examples/blob/master/lib/custom_menu_page.dart) with full examples of how to create custom menus in the meantime.

### CupertinoTextSelectionToolbar
A fully-featured iOS-style toolbar, including overflow handling and positioning.  Since you still must use TextSelectionControls to render this for text selection, there is a bit of extra work to do to get the numbers right (see https://github.com/flutter/flutter/issues/73574).

### CupertinoTextSelectionToolbarButton
A simple wrapper around CupertinoButton to look exactly like the default iOS text selection toolbar button.

### TextSelectionToolbarLayoutDelegate
Positions the toolbar at an anchor, while keeping the menu on the screen.  This was abstracted from the Cupertino and Material versions of this and used by both.

## Related issues
Closes https://github.com/flutter/flutter/issues/36985
Closes https://github.com/flutter/flutter/issues/31412
Closes https://github.com/flutter/flutter/issues/24407

## Tests

  * Tested positioning with TextSelectionToolbarLayoutDelegate.
  * Tested positioning with CupertinoTextSelectionToolbar.
  * Tested overflow with CupertinoTextSelectionToolbar.
  * Tested creating a custom CupertinoTextSelectionToolbar.
  * Tested CupertinoTextSelectionToolbarButton opacity behavior and onPressed.

## Breaking change
Not a breaking change, just exposes new API.

https://user-images.githubusercontent.com/389558/104047661-dd9e1080-5196-11eb-9e7f-e25eaff7c01e.mov

### CODE REVIEW

**Comment by skia-gold** [comment]
Gold has detected about 3 untriaged digest(s) on patchset 19.
View them at https://flutter-gold.skia.org/cl/github/73578

File: packages/flutter/lib/src/widgets/text_selection_toolbar_layout_delegate.dart [hunk 1]
```diff
@@ -0,0 +1,86 @@
+import 'dart:math' as math;
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/text_selection_toolbar_layout_delegate.dart:1]: Is `widgets` the right place for this?  I could also package this as a widget or something.

File: packages/flutter/lib/src/material/text_selection_toolbar.dart [hunk 2]
```diff
@@ -117,66 +131,6 @@ class TextSelectionToolbar extends StatelessWidget {
   }
 }
 
-// Positions the toolbar at the given anchor, ensuring that it remains on
-// screen.
-class _TextSelectionToolbarLayoutDelegate extends SingleChildLayoutDelegate {
```
Comment (justinmc) [inline packages/flutter/lib/src/material/text_selection_toolbar.dart:122]: This has been moved to its own file and made public to be shared here and with CupertinoTextSelectionToolbar. See packages/flutter/lib/src/widgets/text_selection_toolbar_layout_delegate.dart.

File: packages/flutter/lib/src/material/text_selection_toolbar.dart [hunk 3]
```diff
@@ -81,29 +98,26 @@ class TextSelectionToolbar extends StatelessWidget {
...
         + _kToolbarScreenPadding;
-    final double availableHeight = anchorAbove.dy - paddingTop;
+    final double availableHeight = anchorAbove.dy - paddingAbove;
     final bool fitsAbove = _kToolbarHeight <= availableHeight;
-    final Offset anchor = fitsAbove ? anchorAbove : anchorBelow;
-    final Offset localAnchor = Offset(
-      anchor.dx - _kToolbarScreenPadding,
-      anchor.dy - paddingTop,
-    );
+    final Offset localAdjustment = Offset(_kToolbarScreenPadding, paddingAbove);
```
Comment (justinmc) [inline packages/flutter/lib/src/material/text_selection_toolbar.dart:105]: Trying to make things look the same as in the Cupertino version.

File: packages/flutter/lib/src/cupertino/text_selection.dart [hunk 4]
```diff
@@ -196,224 +172,14 @@ class _CupertinoTextSelectionToolbarWrapperState extends State<_CupertinoTextSel
...
-/// An iOS-style toolbar that appears in response to text selection.
-///
-/// Typically displays buttons for text manipulation, e.g. copying and pasting text.
-///
-/// See also:
-///
-///  * [TextSelectionControls.buildToolbar], where [CupertinoTextSelectionToolbar]
-///    will be used to build an iOS-style toolbar.
-@visibleForTesting
-class CupertinoTextSelectionToolbar extends SingleChildRenderObjectWidget {
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_selection.dart:220]: Moved to its own file and made public.

File: packages/flutter/lib/src/cupertino/text_selection.dart [hunk 5]
```diff
@@ -462,45 +229,21 @@ class _CupertinoTextSelectionControls extends TextSelectionControls {
...
-    // since the caller (TextSelectionOverlay._buildToolbar) does not know whether
-    // the toolbar is going to be facing up or down.
-    final double localBarTopY = isArrowPointingDown
-      ? endpoints.first.point.dy - textLineHeight - _kToolbarContentDistance - _kToolbarHeight
-      : endpoints.last.point.dy + _kToolbarContentDistance;
-
-    return _CupertinoTextSelectionToolbarWrapper(
-      arrowTipX: arrowTipX,
-      barTopY: localBarTopY + globalEditableRegion.top,
+    return _CupertinoTextSelectionControlsToolbar(
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_selection.dart:237]: I moved all of the deleted math above into this class here to be consistent with Material.  I'd like to refactor this in the long term (see https://github.com/flutter/flutter/issues/73574), but for now the idea is to encapsulate the complexity and keep the public APIs nice and clean.

File: packages/flutter/lib/src/cupertino/text_selection.dart [hunk 6]
```diff
@@ -570,694 +313,5 @@ class _CupertinoTextSelectionControls extends TextSelectionControls {
   }
 }
 
-// Renders the content of the selection menu and maintains the page state.
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_selection.dart:573]: All of the deleted code here and below was moved to packages/flutter/lib/src/cupertino/text_selection_toolbar.dart and sometimes refactored.

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart [hunk 7]
```diff
@@ -0,0 +1,1028 @@
...
+
+    return Padding(
+      padding: EdgeInsets.fromLTRB(
+        _kToolbarScreenPadding,
+        paddingAbove,
+        _kToolbarScreenPadding,
+        _kToolbarScreenPadding,
+      ),
+      child: CustomSingleChildLayout(
+        delegate: TextSelectionToolbarLayoutDelegate(
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:130]: This logic is now shared with Material via TextSelectionToolbarLayoutDelegate.  I deleted the duplicate logic that did this for Cupertino.

**Comment by skia-gold** [comment]
Gold has detected about 3 untriaged digest(s) on patchset 21.
View them at https://flutter-gold.skia.org/cl/github/73578

File: packages/flutter/lib/src/cupertino/text_selection_toolbar.dart [hunk 8]
```diff
@@ -0,0 +1,1028 @@
...
+      ),
+    );
+  }
+}
+
+// Clips the child so that it has the shape of the default iOS text selection
+// toolbar, with rounded corners and an arrow pointing at the anchor.
+//
+// The anchor should be in global coordinates.
+class _CupertinoTextSelectionToolbarShape extends SingleChildRenderObjectWidget {
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:149]: Most of the rest of the code in this file was copied over from cupertino/text_selection.dart with a little refactoring.  This class in particular was separated out from the positioning logic.

**Comment by skia-gold** [comment]
Gold has detected about 3 untriaged digest(s) on patchset 23.
View them at https://flutter-gold.skia.org/cl/github/73578

**Comment by skia-gold** [comment]
Gold has detected about 6 untriaged digest(s) on patchset 24.
View them at https://flutter-gold.skia.org/cl/github/73578

File: packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    required this.child,
+  }) : super(key: key);
+
+  /// The child of this button.
+  ///
+  /// Usually a [Text] or an [Icon].
+  final Widget child;
+
+  /// Called when this button is pressed.
+  final VoidCallback? onPressed;
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart:39]: should we may this non-nullable? If a menu item is not available shouldn't we take it off from the menu?

File: packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+  /// Returns a [Text] widget in the style of the iOS text selection toolbar
+  /// buttons.
+  ///
+  /// Pass the resulting widget into the [child] parameter when using a
+  /// CupertinoTextSelectionToolbarButton in a CupertinoTextSelectionToolbar.
+  static Text getText(String string, [bool enabled = true]) {
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart:46]: nit: getText sounds like we're retrieving something that has already been created. Maybe turn this into a named constructor.

**Review by LongCatIsLooong — COMMENTED** [review]
LGTM but just to make sure I understand this correctly: 

To add custom buttons one must create a new `TextSelectionControls` whose `buiildToolbar` returns a custom `CupertinoTextSelectionToolbar` with custom children? Is there an easy way to just add a new selection menu item, in addition to whatever control buttons (cut/copy/paste etc.) the system provides?

File: packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart (hunk 9 above)
Comment (justinmc) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart:39]: This is used for the "next" button when on the final page of a menu that overflows.  Here's a terrible screenshot of a custom menu that has it, but it also exists in the native-looking menu:

Screen Shot 2021-01-12 at 1 51 30 PM

File: packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart (hunk 10 above)
Comment (justinmc) [reply packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart]: Thanks, that was a great suggestion.  The code is a lot simpler and users can still easily create text buttons or other things like icon buttons.

**Comment by justinmc** [comment]
@LongCatIsLooong That's right, it's still overly complicated to make simple changes to the text selection toolbar even with this PR (and same for Material).  Here I've attempted to hit the major wins without making any breaking changes.  Long term I'd like to also refactor TextSelectionControls, maybe even get rid of it altogether.  I'm tracking this in https://github.com/flutter/flutter/issues/73574.

**Review by LongCatIsLooong — APPROVED** [review]

**Comment by RAHEYO** [comment]
It shows:
```
Missing concrete implementations of 'TextSelectionControls.buildHandle', 'TextSelectionControls.getHandleAnchor', and 'TextSelectionControls.getHandleSize'.
Try implementing the missing methods, or make the class abstract.dart(non_abstract_class_inherits_abstract_member)
```

When using it...

**Comment by justinmc** [comment]
@RAHEYO I just tried my example of customizing the Cupertino menu on Flutter's master branch and it works for me: https://github.com/justinmc/flutter-text-selection-menu-examples/blob/master/lib/custom_cupertino_menu_page.dart

Can you try that and let me know if it still doesn't work?

**Pull request merged**